File: challenge_17/challenge_17.exs
ones = %{
	1 => "one",
	2 => "two",
	3 => "three",
	4 => "four",
	5 => "five",
	6 => "six",
	7 => "seven",
	8 => "eight",
	9 => "nine",
	10 => "ten",
	11 => "eleven",
	12 => "twelve",
	13 => "thirteen",
	14 => "fourteen",
	15 => "fifteen",
	16 => "sixteen",
	17 => "seventeen",
	18 => "eighteen",
	19 => "nineteen"
}

tens = %{ 
	2 => "twenty",
	3 => "thirty",
	4 => "forty",
	5 => "fifty",
	6 => "sixty",
	7 => "seventy",
	8 => "eighty",
	9 => "ninety"
}

defmodule NumsTo do
	def words(n,_,ones,_) when n < 20 do
		ones[n]
	end

	def words(n,a,ones,tens) when tuple_size(a) == 2 do
		cond do
			Dict.has_key?(ones,elem(a,0)) ->
				ones[n]
			elem(a,0) == "0" ->
				tens[String.to_integer(elem(a,1))]
			true ->
				t = tens[String.to_integer(elem(a,1))]
				o = ones[String.to_integer(elem(a,0))]
				t <> o
		end
	end

	def words(_,a,ones,tens) when tuple_size(a) > 2 do
		cond do
			(elem(a,1) <> elem(a,0)) == "00" ->
				ones[String.to_integer(elem(a,2))] <> "hundred"
			Dict.has_key?(ones,String.to_integer((elem(a,1) <> elem(a,0)))) ->
				o = ones[String.to_integer(elem(a,2))]
				h = ones[String.to_integer((elem(a,1) <> elem(a,0)))]  
				o <> "hundredand" <> h 
			Dict.has_key?(tens,String.to_integer(elem(a,1))) && elem(a,0) == "0" ->
				o = ones[String.to_integer(elem(a,2))]
				t = tens[String.to_integer(elem(a,1))] 
				o <> "hundredand" <> t
			true ->
				h = ones[String.to_integer(elem(a,2))]
				t = tens[String.to_integer(elem(a,1))]
				o = ones[String.to_integer(elem(a,0))]
				h <> "hundredand" <> t <> o
		end
	end

	def str_rev(num) do
		num 
		  |> Integer.to_string() 
		  |> String.split("", trim: true) 
		  |> Enum.reverse()
		  |> List.to_tuple() 
	end
end

c = for n <- (1..999), into: "", do: NumsTo.words(n,NumsTo.str_rev(n),ones,tens)
c = c <> "onethousand"
IO.puts(String.length(c))

File: challenge_17/challenge_17.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

int letter_sum(int i);
static int ones[] 		= {0, 3, 3, 5, 4, 4, 3, 5, 5, 4};
static int teens[]    = {3, 6, 6, 8, 8, 7, 7, 9, 8, 8};
static int tens[] 		= {6, 6, 5, 5, 5, 7, 6, 6}; 
static int hundreds[] = {10, 10, 12, 11, 11, 10, 12, 12, 11};
	
int main(void)
{
	int sum = 11;
	int i, ten, hundred;

	for (i = 1; i < 1000; i++){
		if (i < 10) {
			sum += letter_sum(i);
		} else if (i >= 10 && i < 20) {
			sum += letter_sum(i);
		} else if (i >= 20 && i < 100) {
			sum += letter_sum(i);
		} else {
			ten     = i % 100;
			hundred = (i - ten)/100;

			if (i % 100 != 0) sum += 3;
			sum += letter_sum(ten) + hundreds[hundred-1];
		}
	}

	printf("%d\n", sum);
	return 0;
}
	
int letter_sum(int i) 
{
	int one, ten;
	int result = 0;

	if (i >= 0 && i < 10) {
		result = ones[i];
	} else if (i >= 10 && i < 20) {
		result = teens[i - 10];
	} else {
		one     = i % 10;
		result += ones[one];
		ten     = (i - one)/10;
		result += tens[ten-2];
	}

	return result;
}
